sumo/src/normalize.rs

```diff
@@ -0,0 +1,149 @@
+//! Transforms a `raw::Network` into a `Network` that's easier to reason about.
+
+use std::collections::BTreeMap;
+use std::error::Error;
+
+use abstutil::Timer;
+use geom::{Distance, PolyLine, Pt2D, Ring};
+
+use crate::{raw, Edge, InternalEdge, InternalLane, Junction, Lane, Network};
+
+impl Network {
+    /// Reads a .net.xml file and return the normalized SUMO network.
+    pub fn load(path: &str, timer: &mut Timer) -> Result<Network, Box<dyn Error>> {
+        let raw = raw::Network::parse(path, timer)?;
+        timer.start("normalize");
+        let network = Network::from_raw(raw);
+        timer.stop("normalize");
+        Ok(network)
+    }
+
+    fn from_raw(raw: raw::Network) -> Network {
+        let mut network = Network {
+            location: raw.location,
+            normal_edges: BTreeMap::new(),
+            internal_edges: BTreeMap::new(),
+            junctions: BTreeMap::new(),
+            connections: raw.connections,
+        };
+
+        let types: BTreeMap<String, raw::Type> =
+            raw.types.into_iter().map(|t| (t.id.clone(), t)).collect();
+
+        for junction in raw.junctions {
+            if junction.junction_type == "internal" {
+                continue;
+            }
+            network.junctions.insert(
+                junction.id.clone(),
+                Junction {
+                    pt: junction.pt(),
+                    id: junction.id,
+                    junction_type: junction.junction_type,
+                    incoming_lanes: junction.incoming_lanes,
+                    internal_lanes: junction.internal_lanes,
+                    shape: junction.shape.unwrap(),
+                },
+            );
+        }
+
+        for edge in raw.edges {
+            if edge.function == raw::Function::Internal {
+                let mut lanes = Vec::new();
+                for lane in edge.lanes {
+                    lanes.push(InternalLane {
+                        id: lane.id,
```

Comment:
Ah, so here's where LaneId comes from. So the answer to whether InternalLane.id is a distinct namespace from Lane.id is "whatever SUMO does".

But seems like we treat them as a separate namespace, since we store internal_edges separate from edges.

Reply:
same namespace. internal lane ids always start with a ':'

Reply:
I was able to distinguish the two ID types and choose when looking at `function`.  From what I could tell, a connection's `via` field is always the internal lane ID.